File: problem-solving/_lc-206-easy-reverse-linked-list/2-recursive.py
"""
https://leetcode.com/problems/reverse-linked-list/

Complexity
----------
Time and space: O(n)
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def reverse(prev, curr):
            if curr is None:
                return prev
            newHead = reverse(curr, curr.next)
            curr.next = prev
            return newHead
        return reverse(None, head)

File: problem-solving/lc-206-easy-reverse-linked-list/2-recursive.go
/*
https://leetcode.com/problems/reverse-linked-list/

Complexity
----------
Time and space: O(n)
*/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reverse(prev, curr *ListNode) *ListNode {
    if curr == nil {
        return prev
    }
    newHead := reverse(curr, curr.Next)
    curr.Next = prev
    return newHead
}

func reverseList(head *ListNode) *ListNode {
    return reverse(nil, head)    
}
